### PR TITLE
[NOTEPAD] Fix and improve DoSaveFile

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-18832](https://jira.reactos.org/browse/CORE-18832)

## Proposed changes

- Use `EM_GETHANDLE` to get the text.
- Check zero size and handle it elegantly.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Based on my findings https://jira.reactos.org/browse/CORE-18832?focusedId=138498&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-138498 you do seem to touch the correct module with notepad itself.

**Review by learn-more — APPROVED** [review]

**Comment by JoachimHenze** [comment]
Can you explain with words, where the old code-flow left the expected path before your changes? Thx.

File: base/applications/notepad/dialog.c [hunk 1]
```diff
@@ -332,48 +333,46 @@ GetPrintingRect(HDC hdc, RECT margins)
...
     {
         ShowLastError();
         return FALSE;
     }
 
-    size = GetWindowTextLength(Globals.hEdit) + 1;
-    pTemp = HeapAlloc(GetProcessHeap(), 0, size * sizeof(*pTemp));
-    if (!pTemp)
+    cchText = GetWindowTextLengthW(Globals.hEdit);
+    if (cchText <= 0)
```
Comment (katahiromz) [inline base/applications/notepad/dialog.c:349]: Here, Notepad will check if the text is empty.

Comment (JoachimHenze) [reply base/applications/notepad/dialog.c:349]: Yes, I do see that, thx. But at which point did the **old code** leave the expected path? Thx in advance!

File: base/applications/notepad/dialog.c [hunk 2]
```diff
@@ -332,48 +333,46 @@ GetPrintingRect(HDC hdc, RECT margins)
...
-        ShowLastError();
-        return FALSE;
+        bRet = TRUE;
     }
-    size = GetWindowText(Globals.hEdit, pTemp, size);
-
-    if (size)
+    else
     {
-        if (!WriteText(hFile, (LPWSTR)pTemp, size, Globals.encFile, Globals.iEoln))
```
Comment (katahiromz) [inline base/applications/notepad/dialog.c:360]: The old code fails at `WriteText` due to emptiness.

**Review by JoachimHenze — APPROVED** [review]

File: base/applications/notepad/dialog.c (hunk 2 above)
Comment (katahiromz) [reply base/applications/notepad/dialog.c:360]: https://github.com/reactos/reactos/blob/5913e5bd6b9e41fd3fa77de923c94d8d5d43b92d/base/applications/notepad/text.c#L361

Comment (katahiromz) [reply base/applications/notepad/dialog.c:360]: `iRequiredBytes` was zero for empty files.

**Review by HBelusca — APPROVED** [review]

**Pull request merged**